keps/0021-kudo-upgrade.md

```diff
@@ -0,0 +1,271 @@
+         * [CRDs](#crds)
+            * [Proposal for update process](#proposal-for-update-process-2)
+         * [KUDO CLI](#kudo-cli)
+            * [Proposal for update process](#proposal-for-update-process-3)
+      * [Updating KUDO installation](#updating-kudo-installation)
+         * [Upgrade Steps](#upgrade-steps)
+      * [User Stories](#user-stories)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations TODO](#risks-and-mitigations-todo)
+            * [Failure cases](#failure-cases)
+      * [Graduation Criteria TODO](#graduation-criteria-todo)
+      * [Drawbacks TODO [optional]](#drawbacks-todo-optional)
+      * [Alternatives](#alternatives)
+      * [Infrastructure Needed](#infrastructure-needed)
+      * [Resources](#resources)
+      * [Implementation History](#implementation-history)
+
+<!-- Added by: aneumann, at: Mon Dec  2 11:34:32 CET 2019 -->
+
+<!--te-->
+
+## Summary
+
+TBD
+
+## Motivation
+
+We strive for quick and regular updates of KUDO. We need a process for upgrading all the moving parts of KUDO, and how 
+the different parts interact, and what kind of compatibility we want to provide.
+
+### Goals
+
+- How updates of KUDO are executed
+  - Updates of CRDs
+  - Updates of the KUDO manager
+  - Updates of prerequisites (Namespaces, RoleBindings, ServiceAccounts, etc.)
+- Interoperability
+  - How and if multiple versions of CRDs are maintained
+  - How a version of KUDO CLI work with older and newer CRD versions
+- How to handle operators that are not supported by a new KUDO version
+
+### Non-Goals
+
+- Updates/Upgrades of operators and operator versions itself
+- Versioning of KUDO itself
+- Multi-Tenancy, i.e. running multiple KUDO installations on the same cluster
+
+
+### Current State
+
+At the moment, KUDO does not provide any migration capabilities and needs a clean installation to use a new version.
+
+## Open Questions
+- Lowest supported K8s version
+- Do we want to support downgrades? 
+- Split versions between KUDO manager and KUDO CLI?
+
+## Proposal
+
+### KUDO Prerequisites
+Expected update frequency: Low  
+Versioned: No, but closely tied to KUDO manager
+
+The KUDO manager has a set of prerequisites that need to be present for the manager to run successfully. They are
+the least likely to change, but probably the most specific. If we make changes here, we need to implement custom
+migration code.
+
+- Current prereqs
+  - Namespace
+  - Service Account
+  - Role Bindings
+  - Secret
+
+#### Proposal for update process 
+Integrated into `kudo init --upgrade`
+
+Write specific migration code that targets a KUDO manager version range and executes manual migration steps. 
+
+Each migration should have a validate-step that checks if the migration is possible.
+
+**Alternative update process**
+- Can we just delete them all and reinstall them? Probably not
```

Comment:
what about updating them? just as you do with any other resource, what are the challenges with that?

Reply:
Yes, updating is probably the only possible way. The idea with "Delete all and install fresh" would be to not do migrations but to always just install the expected state, but I don't think that's really possible.